parlai/agents/transformer/polyencoder.py

```diff
@@ -0,0 +1,289 @@
+        bsz = batch.text_vec.size(0)
+        ctxt_rep, ctxt_rep_mask, _ = self.model(ctxt_tokens=batch.text_vec)
+
+        if cand_encs is not None:
+            if bsz == 1:
+                cand_rep = cand_encs
+            else:
+                cand_rep = cand_encs.expand(bsz, cand_encs.size(1), -1)
+        elif len(cand_vecs.shape) == 3:
+            _, _, cand_rep = self.model(cand_tokens=cand_vecs)
+        elif len(cand_vecs.shape) == 2:
+            _, _, cand_rep = self.model(cand_tokens=cand_vecs.unsqueeze(1))
+            cand_rep = cand_rep.expand(bsz, bsz, -1).transpose(0, 1).contiguous()
+        scores = self.model(ctxt_rep=ctxt_rep,
+                            ctxt_rep_mask=ctxt_rep_mask,
+                            cand_rep=cand_rep)
+        return scores
+
+
+class PolyEncoderModule(torch.nn.Module):
+    """ See https://arxiv.org/abs/1905.01969
+    """
+
+    def __init__(self, opt, dict, null_idx):
+        super(PolyEncoderModule, self).__init__()
+        self.null_idx = null_idx
+        self.encoder_ctxt = self.get_encoder(opt, dict, null_idx, 'none')
+        self.encoder_cand = self.get_encoder(opt, dict, null_idx, opt['reduction_type'])
+
+        self.type = opt['polyencoder_type']
+        self.n_codes = opt['poly_n_codes']
+        self.attention_type = opt['poly_attention_type']
+        self.attention_num_heads = opt['poly_attention_num_heads']
+        self.codes_attention_type = opt['codes_attention_type']
+        self.codes_attention_num_heads = opt['codes_attention_num_heads']
+        embed_dim = opt['embedding_size']
+
+        # In case it's a polyencoder with code.
+        if self.type == 'codes':
+            # experimentally it seems that random with size = 1 was good.
+            codes = torch.empty(self.n_codes, embed_dim)
+            codes = torch.nn.init.uniform_(codes)
+            self.codes = torch.nn.Parameter(codes)
+
+            # The attention for the codes.
+            if self.codes_attention_type == 'multihead':
+                self.code_attention = MultiHeadAttention(
+                                        self.codes_attention_num_heads,
+                                        embed_dim,
+                                        opt['dropout'])
+            elif self.codes_attention_type == 'basic_sqrt':
+                self.code_attention = BasicAttention(dim=2, attn='sqrt', get_weights=False)
+            elif self.codes_attention_type == 'basic':
+                self.code_attention = BasicAttention(dim=2, attn='basic', get_weights=False)
+
+        # The final attention (the one that takes the candidate as key)
+        if self.attention_type == 'multihead':
+            self.attention = MultiHeadAttention(
+                                self.attention_num_heads,
+                                opt['embedding_size'],
+                                opt['dropout'])
+        elif self.attention_type == 'basic_sqrt':
```

Comment:
same here

Reply:
Same question as above, I'd rather not change a parameter of BasicAttention (so far I have only added some)

Reply:
Dine